hsm/manager_hsm_test.go

```diff
@@ -0,0 +1,607 @@
+//go:build hsm
+// +build hsm
+
+package hsm_test
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"reflect"
+	"testing"
+
+	"github.com/ThalesIgnite/crypto11"
+	"github.com/golang/mock/gomock"
+	"github.com/miekg/pkcs11"
+	"github.com/pborman/uuid"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/cryptosigner"
+
+	"github.com/ory/hydra/hsm"
+	"github.com/ory/hydra/x"
+)
+
+func TestKeyManager_GenerateKeySet(t *testing.T) {
```

Comment:
As far as I understand, these tests all use mocks. Would it be possible to test the real integration with libhsm as well?

Reply:
`hsm/manager_hsm_test.go` is a unit test that uses mock hsm context. 

https://github.com/ory/hydra/pull/2625/files#diff-309bfb3906e51c2911792e29568a83de4dea61a39e80bbcaee1aa6558b608cb1R15-R21

https://github.com/ory/hydra/blob/0f0eb71d9ecb1fa637feba573b0405e9b2f57624/hsm/hsm_mock_test.go

SoftHSM is used in tests as follows

`make quicktest` executes all tests without SoftHSM and HSM conf disabled.

`make quicktest-hsm` 

- [builds source with hsm build tag ](https://github.com/ory/hydra/pull/2625/files#diff-ae54bef08e3587b28ad8e93eb253a9a5cd9ea6f4251977e35b88dc6b42329e25R18)
- [sets hsm related env variables](https://github.com/ory/hydra/pull/2625/files#diff-ae54bef08e3587b28ad8e93eb253a9a5cd9ea6f4251977e35b88dc6b42329e25R21-R24)
- [initializes token named `hydra` that env variable `HSM_TOKEN_LABEL` references](https://github.com/ory/hydra/pull/2625/files#diff-ae54bef08e3587b28ad8e93eb253a9a5cd9ea6f4251977e35b88dc6b42329e25R26-R28) so that keypairs can be requested/generated correctly
- [and finaly executes tests](https://github.com/ory/hydra/pull/2625/files#diff-ae54bef08e3587b28ad8e93eb253a9a5cd9ea6f4251977e35b88dc6b42329e25R28)

[Also CircleCI test-hsm job executes all tests with SoftHSM and HSM configuration enabled.](https://github.com/ory/hydra/pull/2625/files#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47R78-R113)

There are tests that are skipped when HSM is enabled because keys cannot be added/updated and only generated on HSM
https://github.com/ory/hydra/pull/2625/files#diff-de166095f49463ddaa5fae245ee0cfc49c99a28b69f619b1e1b9301327cd3a82R26-R28

https://github.com/ory/hydra/pull/2625/files#diff-558a3da4e777c67907421d3d7da4d33b537fa8fba7c600b2938bb2d39fdb59faR123-R144

https://github.com/ory/hydra/pull/2625/files#diff-acff70380aa55cc1a224211e29c8c9f530c8db2274f1b57456b793d79d7d9e6cR98-R100

https://github.com/ory/hydra/pull/2625/files#diff-acff70380aa55cc1a224211e29c8c9f530c8db2274f1b57456b793d79d7d9e6cR159-R161

https://github.com/ory/hydra/pull/2625/files#diff-3c4f36e62fc194cb1f350aa6cbbed40cf9407d560478c878d742de3f2a070854R56-R58

Or tests that had to be modified to minimally change the inital test

https://github.com/ory/hydra/pull/2625/files#diff-3c4f36e62fc194cb1f350aa6cbbed40cf9407d560478c878d742de3f2a070854R83-R85

https://github.com/ory/hydra/pull/2625/files#diff-881d683e78d6015e1c6ea6e60c21e442a5dea7e0c7ffc2cacae076e7d498e8d1R76-R79

https://github.com/ory/hydra/pull/2625/files#diff-acff70380aa55cc1a224211e29c8c9f530c8db2274f1b57456b793d79d7d9e6cR57-R59

https://github.com/ory/hydra/pull/2625/files#diff-acff70380aa55cc1a224211e29c8c9f530c8db2274f1b57456b793d79d7d9e6cR70-R75

https://github.com/ory/hydra/pull/2625/files#diff-acff70380aa55cc1a224211e29c8c9f530c8db2274f1b57456b793d79d7d9e6cR129-R133

https://github.com/ory/hydra/pull/2625/files#diff-acff70380aa55cc1a224211e29c8c9f530c8db2274f1b57456b793d79d7d9e6cR145-R149